Revise Jörmungandr Binary Fragment Serializers

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

Another take on #1021 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have revised Jörmungandr binary serializers for fragments. Goal is to stay close to the grammar and avoid needless indirections, while still being flexible enough to be used to implement what we have to implement.


# Comments

<!-- Additional comments or screenshots to attach if any -->

```
decode (encode fragment) == fragment
  +++ OK, passed 1000 tests:
  50.9% Transaction
  25.2% StakeDelegation (None)
  23.9% StakeDelegation (Full)

decodeExternalTx works ok with properly constructed binary blob
  +++ OK, passed 100 tests:
  59% Simple Transaction
  41% Not Simple Transaction
```

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
revise Jörmungandr binary serializers for fragment (normal transaction and delegation certs)

KtorZ committed:
propagate the changes to Jörmungandr transaction layer

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -537,42 +460,117 @@ getLegacyTransaction tid = do
+        putWord8 1
+        putByteString $ unXSignature $ sign pwd xprv accountId
+
+sealFragment :: Put -> (Hash "Tx", SealedTx)
```
I originally contemplated naming this `seal` too. I like it. But in `jcli` `seal` comes _before_ `auth`, e.g. in:
```
jcli transaction new | jcli transaction add-certificate (jcli utils bech32-convert (cat deleg.cert) cert) | jcli transaction finalize | jcli transaction seal | jcli transaction auth -k key.prv | jcli transaction to-message
```

So we may have a different notion of `seal`.

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -537,42 +460,117 @@ getLegacyTransaction tid = do
+    = MkFragmentSimpleTransaction
+        TxWitnessTag
+    | MkFragmentStakeDelegation
+        TxWitnessTag
```
As I've said previously, this restricts us to to have transactions with a **single type** of inputs / witnesses.

No — we don't need it _at the moment_, but I think some worry is warranted when making these kinds of big abstractions.

I see that if we need it, we could move it to the list of inputs, so we should probably be very fine though 👍

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -738,6 +736,15 @@ putAddress (Address bs) = putByteString bs
 -------------------------------------------------------------------------------}
 
+guardLength :: (Int -> Bool) -> [a] -> Put
+guardLength predicate xs = do
```
👍

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -537,42 +460,117 @@ getLegacyTransaction tid = do
+        putByteString $ unXSignature $ sign pwd xprv accountId
+
+sealFragment :: Put -> (Hash "Tx", SealedTx)
+sealFragment put = (Hash (blake2b256 bytes), SealedTx bytes)
```
Careful here! I think the `fragmentId` might exclude the length part of the fragment (I know I said "all" in a comment but…)

KtorZ replied:
Indeed, good catch! Property test will tell too :)

KtorZ replied:
I think you're right :thinking:

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -537,42 +460,117 @@ getLegacyTransaction tid = do
+      where
+        msg = mconcat
+            [ block0Hash
+            , blake2b256 $ BL.toStrict $ runPut putInputsOutputs
```
As I've said already, I'm pretty sure this should be `blake2b256 (payload <> putInputsOutputs)`

KtorZ replied:
We currently have: 

```hs
            let payload = block0H <> getHash (signData inps outs)
```

Where

```hs
signData
    :: [(TxIn, Coin)]
    -> [TxOut]
    -> Hash "SignData"
signData inps outs =
    Hash $ blake2b256 $ BL.toStrict $ runPut $ do
        putTx inps outs
```

So I think this is correct, and clearer at the same time.

Anviking replied:
I meant the payload as in where the certificate goes.

Anviking reviewed (commented):
I think there are still mistakes that have the potential to make your abstractions less nice.

The comments from the spec are of course nice. But I don't really see why this had to be a complete remake of #1059.

KtorZ committed:
revise Jörmungandr unit tests, in particular binary roundtrip for fragments

Now combined both serialization and deserialization for transaction and stake delegation into one property test
with proper labelling, shrinking and counter examples. Also removed all the wrong exports of the binary internals
onto the test and solely rely on the top-level 'getFragment' and 'putFragment' functions from the module

KtorZ committed:
Move decode external tx tests from integation to unit (Binary module)

Aaaaand, revise them to also use 'putFragment'

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -537,42 +461,122 @@ getLegacyTransaction tid = do
+sealFragment put =
+    -- NOTE:
+    -- The fragment id is a hash of the fragment, minus the size (first 2 bytes)
+    (Hash (blake2b256 $ BS.drop 2 bytes), SealedTx bytes)
```
😕

KtorZ replied:
Yeah. Not so happy either. That's provisional, hopefully. I am still working on getting the integration tests to run.

KtorZ committed:
revise NetworkSpec integration tests to use 'putFragment'

KtorZ committed:
revise tests about max transaction size

This is no longer a thing on Jörmungandr. The number of inputs are outputs are limited, but
apart from that, transactions seems not to have any maximum size

KtorZ committed:
add provisional golden test for stake delegation tx

KtorZ committed:
Attempt to make top-level API a bit more type-safe regarding fragment creation

Anviking commented on lib/core/src/Cardano/Wallet/Unsafe.hs:
```diff
@@ -74,8 +74,8 @@ unsafeFromText = either (error . show) id . fromText
 
 -- | Build a 'XPrv' from an hex-encoded bytestring
```
Out-of-date comment?

Anviking replied:
```suggestion
 -- | Build a 'XPrv' from a bytestring
```

KtorZ replied:
Indeed!

Anviking replied:
Oops, the suggested change diff lied
<img width="632" alt="Skärmavbild 2019-11-25 kl  12 44 28" src="https://user-images.githubusercontent.com/304423/69537895-52a58b00-0f81-11ea-8b64-f8b98f92fb2d.png">

where as the real change added a space before the comment 🤷‍♂

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs:
```diff
@@ -63,39 +58,35 @@ import Data.Text.Class
+                (MkFragmentSimpleTransaction (txWitnessTagFor @k))
+        return
+            ( Tx
+                { txId = fragmentId fragment
```
We _could_ (like in https://github.com/input-output-hk/cardano-wallet/pull/1021#discussion_r349290864) let `putFragment` return `Hash "Tx", SealedTx` also, and we'd
1. Be able to remove `fragmentId`
2. No longer have to call `BS.drop 2`

Anviking replied:
On that topic, I think we're missing tests for `fragmentId`. (not necessarily because of this PR)

We should test that `fragmentId` here and `getFragment` results in the same id, and that it matches `jcli`.

KtorZ replied:
> We should test that fragmentId here and getFragment results in the same id, and that it matches jcli.

I agree. Currently, if all our unit tests pass, we have no guarantee that ids are computed correctly, which is kinda problematic.

KtorZ replied:
> We could (like in #1021 (comment)) let putFragment return Hash "Tx", SealedTx also, and we'd
> 
>    Be able to remove fragmentId
>    No longer have to call BS.drop 2

I strongly disagree here. `putFragment` should return a `PutM` and nothing else. Returning something else makes it harder to test and less intuitive. 

Being able to split out `fragmentId` into a separate function is actually a good thing (single responsibility principle for function, more composable). 

> No longer have to call BS.drop 2

Sort of inelegant, but very much acceptable in the binary module IMO.

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -537,42 +468,128 @@ getLegacyTransaction tid = do
+        withHeader FragmentDelegation $ do
+            putPayload  <- returnPut $ putDelegationCertificate account dlgTag
+            putAuthData <- returnPut $ putTransaction putPayload witTag
+            putStakeAuthentication (putPayload >> putAuthData) credentials
```
Just for the record, what I liked about #1021 was that [this structure](https://github.com/input-output-hk/cardano-wallet/pull/1059/files#diff-8e20179d01c474d1713ed389da590cd8R590-R593) was strictly/automatically enforced such that we wouldn't have to pass in `(putPayload >> putAuthData)` here (for each case that requires it).

But it works :+1:

Commit:
Adjust outdated comment in the unsafe module

Co-Authored-By: anviking <johannes.lund@iohk.io>
